File: Core/State/Automatic/Extensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BinkyRailways.Core.Model;

namespace BinkyRailways.Core.State.Automatic
{
    internal static class Extensions
    {
        /// <summary>
        /// Is is needed for a loc facing in the given direction (in the From block of the given route)
        /// to chance direction when taking this route?
        /// </summary>
        /// <param name="route">The route being investigated</param>
        /// <param name="locDirection">The direction the loc is facing in the From block of the given <see cref="route"/>.</param>
        internal static bool IsDirectionChangeNeeded(this IRouteState route, BlockSide locDirection)
        {
            var fromSide = route.FromBlockSide;
            return (fromSide != locDirection);
        }

        /// <summary>
        /// Gets a selection of all possible routes from the given block.
        /// </summary>
        internal static IEnumerable<IRouteState> GetAllPossibleRoutesFromBlock(this IRailwayState railwayState, IBlockState fromBlock)
        {
            return railwayState.RouteStates.Where(x => (x.From == fromBlock));
        }

        /// <summary>
        /// Gets a selection of all possible non-closed routes from the given block.
        /// </summary>
        internal static IEnumerable<IRouteState> GetAllPossibleNonClosedRoutesFromBlock(this IRailwayState railwayState, IBlockState fromBlock)
        {
            return railwayState.GetAllPossibleRoutesFromBlock(fromBlock).Where(x => !x.Closed);
        }

        /// <summary>
        /// Gets a selection of all possible routes from the given block + side.
        /// </summary>
        internal static IEnumerable<IRouteState> GetAllPossibleRoutesFromBlock(this IRailwayState railwayState, IBlockState fromBlock, BlockSide fromBlockSide)
        {
            return railwayState.RouteStates.Where(x => (x.From == fromBlock) && (x.FromBlockSide == fromBlockSide));
        }

        /// <summary>
        /// Gets a selection of all possible non-closed routes from the given block + side.
        /// </summary>
        internal static IEnumerable<IRouteState> GetAllPossibleNonClosedRoutesFromBlock(this IRailwayState railwayState, IBlockState fromBlock, BlockSide fromBlockSide)
        {
            return railwayState.GetAllPossibleRoutesFromBlock(fromBlock, fromBlockSide).Where(x => !x.Closed);
        }

        /// <summary>
        /// Are there more than 1 non-closed routes from the given block.
        /// </summary>
        internal static bool HasMultipleNonClosedRoutesFromBlock(this IRailwayState railwayState, IBlockState fromBlock)
        {
            return (railwayState.GetAllPossibleNonClosedRoutesFromBlock(fromBlock).Count() > 1);
        }

        /// <summary>
        /// Can the given loc change direction in the given block?
        /// </summary>
        internal static bool CanChangeDirectionIn(this ILocState loc, IBlockState block)
        {
            return (loc.ChangeDirection == ChangeDirection.Allow) && (block.ChangeDirection == ChangeDirection.Allow);
        }

        /// <summary>
        /// Is the given block consider a station for the given loc?
        /// </summary>
        /// <remarks>It is considered a station if the loc is allowed to stop in the block and the wait possibility is greater then 50%</remarks>
        internal static bool IsStationFor(this IBlockState block, ILocState loc)
        {
            return (block.IsStation) && block.WaitPermissions.Evaluate(loc);
        }

        /// <summary>
        /// Take a gamble.
        /// The result is true or false, where the given chance is for "true".
        /// </summary>
        internal static bool Gamble(this Random rnd, int chance)
        {
            if ((chance < 0) || (chance > 100))
                throw new ArgumentOutOfRangeException("chance", chance, "Must be between 0..100");
            if (chance == 0)
                return false;
            if (chance == 100)
                return true;
            var value = rnd.Next(100);
            return (value <= chance);
        }

        /// <summary>
        /// Take a random choice between one of the options.
        /// Each option has a probability.
        /// </summary>
        internal static T Gamble<T>(this Random rnd, params Tuple<T, int>[] options)
        {
            var boundaries = new int[options.Length];
            var max = 0;
            for (var i = 0; i < options.Length; i++)
            {
                max += options[i].Item2 + 1;
                boundaries[i] = max;
            }

            var value = rnd.Next(max + 1);
            for (var i = 0; i < options.Length; i++)
            {
                if (value <= boundaries[i])
                    return options[i].Item1;
            }
            // We should never get here
            return options[options.Length - 1].Item1;
        }

        /// <summary>
        /// Is the given loc allowed to leave its current block?
        /// </summary>
        internal static bool CanLeaveCurrentBlock(this ILocState loc)
        {
            var currentBlock = loc.CurrentBlock.Actual;
            var currentBlockGroup = (currentBlock != null) ? currentBlock.BlockGroup : null;
            return ((currentBlockGroup == null) || (currentBlockGroup.FirstLocCanLeave));
        }
    }
}

File: pkg/core/state/automatic/extensions.go
// Copyright 2022 Ewout Prangsma
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author Ewout Prangsma
//

package automatic

import (
	"context"
	"math/rand"
	"time"

	"github.com/binkyrailways/BinkyRailways/pkg/core/model"
	"github.com/binkyrailways/BinkyRailways/pkg/core/state"
)

func init() {
	rand.Seed(time.Now().Unix())
}

// Is is needed for a loc facing in the given direction (in the From block of the given route)
// to chance direction when taking this route?
// route: The route being investigated
// locDirection: The direction the loc is facing in the From block of the given <see cref="route"/>.
func isDirectionChangeNeeded(ctx context.Context, route state.Route, locDirection model.BlockSide) bool {
	fromSide := route.GetFromBlockSide()
	return (fromSide != locDirection)
}

// Is the given loc allowed to leave its current block?
func canLeaveCurrentBlock(ctx context.Context, loc state.Loc) bool {
	currentBlock := loc.GetCurrentBlock().GetActual(ctx)
	if currentBlock == nil {
		return true
	}
	currentBlockGroup := currentBlock.GetBlockGroup(ctx)
	return ((currentBlockGroup == nil) || (currentBlockGroup.GetFirstLocCanLeave(ctx)))
}

// Can the given loc change direction in the given block?
func canChangeDirectionIn(ctx context.Context, loc state.Loc, block state.Block) bool {
	return (loc.GetChangeDirection(ctx) == model.ChangeDirectionAllow) &&
		(block.GetChangeDirection(ctx) == model.ChangeDirectionAllow)
}

// Is the given block consider a station for the given loc?
// Note: It is considered a station if the loc is allowed to stop in the block and the wait possibility is greater then 50%.
func isStationFor(ctx context.Context, block state.Block, loc state.Loc) bool {
	return (block.GetIsStation(ctx)) && block.GetWaitPermissions().Evaluate(ctx, loc)
}

// Take a gamble.
// The result is true or false, where the given chance is for "true".
func gamble(percent int) bool {
	//if ((chance < 0) || (chance > 100))
	//    throw new ArgumentOutOfRangeException("chance", chance, "Must be between 0..100");
	if percent == 0 {
		return false
	}
	if percent == 100 {
		return true
	}
	value := rand.Intn(100)
	return (value <= percent)
}
